ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/DataColumnSidecarRecoveringCustodyImpl.java

```diff
@@ -0,0 +1,290 @@
+                      .toList();
+              ids.stream()
+                  .map(recoveryTasks::get)
+                  .forEach(
+                      task -> {
+                        task.timeouted().set(true);
+                        maybeStartRecovery(task);
+                      });
+            },
+            slotToRecoveryDelay.apply(slot))
+        .ifExceptionGetsHereRaiseABug();
+  }
+
+  @Override
+  public void onNewBlock(
+      final SignedBeaconBlock block, final Optional<BlobSidecarManager.RemoteOrigin> remoteOrigin) {
+    if (!isActiveSuperNode(block.getSlot())) {
+      return;
+    }
+
+    if (remoteOrigin.isPresent()
+        && (remoteOrigin.get().equals(BlobSidecarManager.RemoteOrigin.LOCAL_EL)
+            || remoteOrigin.get().equals(BlobSidecarManager.RemoteOrigin.LOCAL_PROPOSAL))) {
+      // skip locally produced blocks, we will get everything for it in custody w/o reconstruction
+      return;
+    }
+    createOrUpdateRecoveryTaskForBlock(block.getMessage());
+  }
+
+  private synchronized void createOrUpdateRecoveryTaskForBlock(final BeaconBlock block) {
+    if (recoveryTasks.containsKey(block.getSlotAndBlockRoot())) {
+      final RecoveryTask existing = recoveryTasks.get(block.getSlotAndBlockRoot());
+      if (existing.block().get() == null) {
+        existing.block().set(block);
+        maybeStartRecovery(existing);
+      }
+    } else {
+      recoveryTasks.put(
+          block.getSlotAndBlockRoot(),
+          new RecoveryTask(
+              new AtomicReference<>(block),
+              new HashSet<>(),
+              new AtomicBoolean(false),
+              new AtomicBoolean(false)));
+    }
+  }
+
+  private synchronized void maybeStartRecovery(final RecoveryTask task) {
+    if (readyToBeRecovered(task)) {
+      task.recoveryStarted().set(true);
+      if (task.existing().size() != columnCount) {
+        asyncRunner.runAsync(() -> prepareAndInitiateRecovery(task)).ifExceptionGetsHereRaiseABug();
+      }
+    }
+  }
+
+  private boolean readyToBeRecovered(final RecoveryTask task) {
+    if (task.recoveryStarted().get()) {
+      // already started
+      return false;
+    }
+
+    if (!task.timeouted().get()) {
+      return false;
+    }
+
+    if (task.existing().size() < recoverColumnCount) {
+      // not enough columns collected
+      return false;
+    }
+
+    if (task.block().get() == null) {
+      return false;
+    }
+
+    return true;
+  }
+
+  @Override
+  public void subscribeToValidDataColumnSidecars(
+      final DataColumnSidecarManager.ValidDataColumnSidecarsListener sidecarsListener) {
+    validDataColumnSidecarsSubscribers.subscribe(sidecarsListener);
+  }
+
+  private boolean isActiveSuperNode(final UInt64 slot) {
+    return isSuperNode
+        && spec.atSlot(slot).getMilestone().isGreaterThanOrEqualTo(SpecMilestone.FULU);
+  }
+
+  private record RecoveryTask(
+      AtomicReference<BeaconBlock> block,
+      Set<DataColumnSlotAndIdentifier> existing,
```

Comment:
nit: maybe more descriptive field name for better code reading: `existingColumnIds` e.g.?